qiskit_machine_learning/algorithms/trainable_model.py

```diff
@@ -0,0 +1,160 @@
+"""A base ML model with a Scikit-Learn like interface."""
+
+from abc import abstractmethod
+from typing import Union, Optional
+
+import numpy as np
+from qiskit.algorithms.optimizers import Optimizer
+
+from qiskit_machine_learning import QiskitMachineLearningError
+from qiskit_machine_learning.neural_networks import NeuralNetwork
+from qiskit_machine_learning.utils.loss_functions import (
+    Loss,
+    L1Loss,
+    L2Loss,
+    CrossEntropyLoss,
+    CrossEntropySigmoidLoss,
+)
+
+
+class TrainableModel:
+    """Base class for ML model. This class defines Scikit-Learn like interface to implement."""
+
+    def __init__(
+        self,
+        neural_network: NeuralNetwork,
+        loss: Union[str, Loss] = "l2",
+        optimizer: Optimizer = None,
+        warm_start: bool = False,
+    ):
+        """
+        Args:
+            neural_network: An instance of an quantum neural network. If the neural network has a
+                one-dimensional output, i.e., `neural_network.output_shape=(1,)`, then it is
+                expected to return values in [-1, +1] and it can only be used for binary
+                classification. If the output is multi-dimensional, it is assumed that the result
+                is a probability distribution, i.e., that the entries are non-negative and sum up
+                to one. Then there are two options, either one-hot encoding or not. In case of
+                one-hot encoding, each probability vector resulting a neural network is considered
+                as one sample and the loss function is applied to the whole vector. Otherwise, each
+                entry of the probability vector is considered as an individual sample and the loss
+                function is applied to the index and weighted with the corresponding probability.
+            loss: A target loss function to be used in training. Default is `l2`, i.e. L2 loss.
+                Can be given either as a string for 'l1', 'l2', 'cross_entropy',
+                'cross_entropy_sigmoid', or as a loss function implementing the Loss interface.
+            optimizer: An instance of an optimizer to be used in training.
+            warm_start: Use weights from previous fit to start next fit.
+
+        Raises:
+            QiskitMachineLearningError: unknown loss, invalid neural network
+        """
+        self._neural_network = neural_network
+        if len(neural_network.output_shape) > 1:
+            raise QiskitMachineLearningError("Invalid neural network output shape!")
+        if isinstance(loss, Loss):
+            self._loss = loss
+        else:
+            loss = loss.lower()
+            if loss == "l1":
+                self._loss = L1Loss()
+            elif loss == "l2":
+                self._loss = L2Loss()
+            elif loss == "cross_entropy":
+                self._loss = CrossEntropyLoss()
+            elif loss == "cross_entropy_sigmoid":
+                self._loss = CrossEntropySigmoidLoss()
```

Comment:
Shall we extract it to a separate class that would handle providing a correct Loss object based on the string? I can imagine that this string-based convention might reappear in many classes and then each of them needs this long block of if statements which leads to code duplication.

Reply:
In general, I agree. But I'd rather wait till the time this situation occurs or address it when #82 will be in progress.